### PR TITLE
Support for 'docker-compose.override.yaml' and 'docker-compose -f <file>'

### PR DESCRIPTION
**Pull request by aschrijver**
I added support for having a default override `docker-compose.override.yml` and to explicitly specify the docker-compose files to use. For this last bit I added a `useComposeFiles` config option.

### CODE REVIEW

**Comment by aschrijver** [comment]
Looks like Travis CI is not properly initializing. It cannot install docker-engine.

**Comment by Augi** [comment]
I fixed the Docker Engine installation - it's in `master` now so you can merge it to your branch.

I will look at the MR itself later today. Thanks for contribution!

**Comment by aschrijver** [comment]
Everything seems to be in working order now. Thank you for the plugin!  :)

File: src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeDown.groovy [hunk 1]
```diff
@@ -37,10 +37,17 @@ class ComposeDown extends DefaultTask {
...
             project.exec { ExecSpec e ->
-                e.commandLine 'docker-compose', '--version'
+                e.commandLine prepareCommand(['docker-compose', '--version'])
                 e.standardOutput = os
             }
             VersionNumber.parse(os.toString().trim().findAll(/(\d+\.){2}(\d+)/).head())
         }
     }
+
+    protected Iterable<String> prepareCommand(List<String> baseCommand) {
```
Comment (Augi) [inline src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeDown.groovy]: Is it required to change commands in `ComposeDown`? I think that `docker-compose` _connects_ particular commands on `project` base - it's name of directory (and can be set by `-p` argument). So I think that we don't have to change `ComposeDown`, we should change `build` and `up` calls only.

Could you please verify this assumption?

File: src/main/groovy/com/avast/gradle/dockercompose/ComposeExtension.groovy [hunk 2]
```diff
@@ -18,6 +18,7 @@ class ComposeExtension {
     boolean waitForTcpPorts = true
     Duration waitAfterTcpProbeFailure = Duration.ofSeconds(1)
     Duration waitForTcpPortsTimeout = Duration.ofMinutes(15)
+    List<String> useComposeFiles
```
Comment (Augi) [inline src/main/groovy/com/avast/gradle/dockercompose/ComposeExtension.groovy]: I don't like `null` handling so I would prefer to initialize `useComposeFiles` field to empty list.

File: src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeUp.groovy [hunk 3]
```diff
@@ -14,6 +14,8 @@ import java.time.Instant
 class ComposeUp extends DefaultTask {
 
     private Map<String, ServiceInfo> servicesInfos = new HashMap<>()
+    private List<String> composeFiles
```
Comment (Augi) [inline src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeUp.groovy]: I would prefer to work with `extension.useComposeFiles` property directly (so remove this field).

File: src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeUp.groovy [hunk 4]
```diff
@@ -14,6 +14,8 @@ import java.time.Instant
 class ComposeUp extends DefaultTask {
 
     private Map<String, ServiceInfo> servicesInfos = new HashMap<>()
+    private List<String> composeFiles
+    private boolean filesSpecified
```
Comment (Augi) [inline src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeUp.groovy]: I think that `extension.useComposeFiles` would be fine, so we don't need this field (if I didn't overlook something).

File: src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeUp.groovy [hunk 5]
```diff
@@ -48,11 +52,28 @@ class ComposeUp extends DefaultTask {
         }
     }
 
+    protected void initializeComposeFilesInfo() {
```
Comment (Augi) [inline src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeUp.groovy]: As both fields would be removed, we could remove this method.

File: src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeUp.groovy [hunk 6]
```diff
@@ -48,11 +52,28 @@ class ComposeUp extends DefaultTask {
         }
     }
 
+    protected void initializeComposeFilesInfo() {
+        composeFiles = extension.useComposeFiles
+        filesSpecified = composeFiles && composeFiles.size() > 0
+    }
+
+    protected void honourComposeFileOverride() {
```
Comment (Augi) [inline src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeUp.groovy]: Why is this needed? According to [the documentation](https://docs.docker.com/compose/reference/overview/), it will look for `docker-compose.override.yml` as well. I would prefer to do nothing if compose files are not specified.

File: src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeUp.groovy [hunk 7]
```diff
@@ -65,10 +86,19 @@ class ComposeUp extends DefaultTask {
...
+            // if there is 'version: 2' on top-level then information about services is in 'services' sub-tree
+            '2'.equals(compose.get('version')) ? ((Map) compose.get('services')).keySet() : compose.keySet()
+        }.unique()
+    }
+
     String getContainerId(String serviceName) {
         new ByteArrayOutputStream().withStream { os ->
             project.exec { ExecSpec e ->
-                e.commandLine 'docker-compose', 'ps', '-q', serviceName
+                e.commandLine prepareCommand(['docker-compose', 'ps', '-q', serviceName])
```
Comment (Augi) [inline src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeUp.groovy]: The `prepareCommand` is not required here IMHO.

File: src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeDown.groovy (hunk 1 above)
Comment (aschrijver) [inline src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeDown.groovy]: Yes, it is required, otherwise you'll get error "_Can't find a suitable configuration file in this directory or any parent. Are you in the right directory?_". When using the -f option you can specify any filename and are not bound by 'docker-compose.yml' standard.

File: src/main/groovy/com/avast/gradle/dockercompose/ComposeExtension.groovy (hunk 2 above)
Comment (aschrijver) [inline src/main/groovy/com/avast/gradle/dockercompose/ComposeExtension.groovy]: Done.

File: src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeUp.groovy (hunk 3 above)
Comment (aschrijver) [inline src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeUp.groovy]: I use this local copy for the case where there is a `docker-compose.override.yml`. In this case a plain `docker-compose up -d` (without -f), but then both yaml's are added to `composeFiles` so service info collection does include the override.
While `composeFiles` changes the `extension.useComposeFiles` does not. It reflects the user's preference.

File: src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeUp.groovy (hunk 6 above)
Comment (aschrijver) [inline src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeUp.groovy]: See my earlier comment on `composeFiles`.
The `filesSpecified` is a shorthand because it is used multiple times.

Also `docker-compose up -d` behaves differently than `docker-compose -f docker-compose.yml up -d`. In the latter case the docker-compose.override.yml is ignored

I moved the code in `honourComposeFileOverride` to `loadServicesInfo` (though can also be in `getServiceNames`.

File: src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeUp.groovy (hunk 7 above)
Comment (aschrijver) [inline src/main/groovy/com/avast/gradle/dockercompose/tasks/ComposeUp.groovy]: Here also you'll get an error if it is missing.

**Comment by Augi** [comment]
Thank you for explanation and contribution, merging. It will be released immediately - check the version on Bintray (click on the icon on README page).

**Comment by aschrijver** [comment]
Great! Thanks a lot!!

I may submit more pull requests later, but will then first create an enhancement issue for it. I am thinking of support for `docker compose up <service...>` first.

**Comment by Augi** [comment]
Btw. I refactored the code little bit. All tests are passing so it should be OK but let me know if there's something strange :-)

**Comment by aschrijver** [comment]
Nicely done! You can tell I'm more of a Java guy still :)
I just refactored my build quite a bit, making it nice and clean, and the latest version is still running fine.
Thanks!
